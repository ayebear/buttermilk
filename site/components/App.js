import React, { Component } from 'react';
import styled from 'styled-components';

import Home from './Home';

class App extends Component {
  render() {
    return (
      <Container>
        <Header>
          <Headline>
            Buttermilk
            <Version href="https://github.com/probablyup/buttermilk/releases" target="_blank">
              {VERSION.slice(0, VERSION.lastIndexOf('.'))}
            </Version>
          </Headline>
        </Header>

        <Content>
          {this.props.children}
        </Content>
      </Container>
    );
  }
}

const desktop = '@media all and (min-width: 768px)';

const Container = styled.div`
  background: #FAFAFA;
  min-height: 100%;
`;

const Header = styled.header`
  align-items: center;
  background: white;
  display: flex;
  flex-direction: column;
  justify-content: center;
  min-height: 125px;

  ${desktop} {
    min-height: 300px;
  }
`;

const Headline = styled.h1`
  font-family: 'Vibur', sans-serif;
  font-size: 5rem;
  margin: 0;
  position: relative;
  transform: rotate(-6deg);
  transform-origin: center;

  ${desktop} {
    font-size: 10rem;
    text-shadow: 0 0 300px white;
    transition: 500ms text-shadow;

    &:hover {
      text-shadow: 0 0 300px;
    }
  }
`;

const Version = styled.a`
  font-size: 0.25em;
  position: absolute;
  right: -1em;
  bottom: -0.5em;
  text-decoration: none;
`;

const Content = styled.main`
  font-size: 1.6rem;
`;

export default () => (
  <App>
    <Home />
  </App>
);
