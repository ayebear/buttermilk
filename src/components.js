import createContext from 'create-react-context';
import PropTypes from 'prop-types';
import React from 'react';

import regexify from './regexify';

import {
    extractParamsFromPath,
    findRoute,
    getDisplayName,
    getRouteParamsForURL,
    parseUrl,
    route,
    valid,
} from './utils';

const CONTEXT = createContext('buttermilk');
const BROWSER = typeof window !== 'undefined';
const NOOP = function () {};

/**
 * The gist of Buttermilk's router is that it acts like a controlled component when used
 * server-side (driven by `props.url`) and an uncontrolled one client-side (driven by the
 * value of `window.location.href` and intercepted navigation events.)
 *
 * In the browser, use either a <Link> component or the route() method to change routes.
 * The router will also automatically pick up popstate events caused by user-driven browser
 * navigation (forward, back buttons, etc.)
 */
export class Router extends React.Component {
    static propTypes = {
        /**
         * Provide a spinner or something to look at while the promise
         * is in flight if using async routes.
         */
        loadingComponent: PropTypes.oneOfType([
            PropTypes.func,
            PropTypes.string,
        ]),

        /**
         * An optional app runtime component. Think of it like the "shell" of your
         * app, so perhaps the outer container, nav bar, etc. You'll probably want to
         * put any "Provider" type components here that are intended to wrap your
         * whole application.
         */
        outerComponent: PropTypes.oneOfType([
            PropTypes.func,
            PropTypes.string,
        ]),

        routes: PropTypes.arrayOf(
            PropTypes.shape({
                /**
                 * A RegExp, string, or function accepting the URL as
                 * an argument and returning a boolean if valid.
                 */
                path: PropTypes.oneOfType([
                    PropTypes.instanceOf(RegExp),
                    PropTypes.string,
                    PropTypes.func,
                ]).isRequired,

                /**
                 * A string URL path to a different route. If this is given,
                 * then "render" is not required.
                 */
                redirect: PropTypes.string,

                /**
                 * A function that returns one of the following:
                 *
                 * 1. JSX.
                 * 2. A React component class.
                 * 3. A promise resolving to JSX or a React component class.
                 */
                render: PropTypes.func,
            }),
        ).isRequired,

        /**
         * A hook for reacting to an impending route transition. Accepts a promise
         * and will pause the route transition until the promise is resolved. Return
         * false or reject a given promise to abort the routing update.
         *
         * Provides currentRouting and nextRouting as arguments.
         */
        routeWillChange: PropTypes.func,

        /**
         * A hook for reacting to a completed route transition. It might be used
         * for synchronizing some global state if desired.
         *
         * Provides currentRouting and previousRouting as arguments.
         */
        routeDidChange: PropTypes.func,

        /**
         * A hook for synchronizing initial routing state.
         *
         * Providers initialRouting as an argument.
         */
        routerDidInitialize: PropTypes.func,

        /**
         * The initial URL to be used for processing, falls back to
         * window.location.href for non-SSR. Required for environments without
         * browser navigation eventing.
         */
        url: PropTypes.string,
    };

    static defaultProps = {
        loadingComponent: 'div',
        outerComponent: 'div',
        routeDidChange: NOOP,
        routeWillChange: NOOP,
        routerDidInitialize: NOOP,
        url: '',
    };

    static getURL = function () { return window.location.href; }

    noFallbackWarningEmitted = false;
    promise = null;

    constructor(props, context) {
        super(props, context);

        if (!BROWSER && !props.url) {
            throw new Error('props.url is required for non-browser environments');
        }

        const routes = this.processRoutes(props.routes);
        const url = props.url || Router.getURL();

        this.state = this.getStateUpdateForUrl(url, routes);

        props.routerDidInitialize(this.getContextValue(url, this.state.activeRoute));
    }

    /**
     * In a browser setting, we only want to rely on browser navigation events
     * to determine routing updates.
     */
    shouldRecompute(nextProps) {
        return      nextProps.routes !== this.props.routes
               || ((nextProps.url !== this.state.url) && !BROWSER);
    }

    componentWillReceiveProps(nextProps) {
        if (this.shouldRecompute(nextProps)) {
            this.recomputeRoutingState(
                BROWSER ? this.state.url : nextProps.url, nextProps.routes
            );
        }
    }

    componentDidMount() {
        window.addEventListener('popstate', this.handleLocationChange);
        window.addEventListener('hashchange', this.handleLocationChange);
    }

    componentWillUnmount() {
        window.removeEventListener('popstate', this.handleLocationChange);
        window.removeEventListener('hashchange', this.handleLocationChange);
    }

    render() {
        const contextValue = this.getContextValue(this.state.url, this.state.activeRoute);

        return (
            <CONTEXT.Provider value={contextValue}>
                <this.props.outerComponent {...contextValue}>
                    {this.renderChildren(this.state.children, contextValue)}
                </this.props.outerComponent>
            </CONTEXT.Provider>
        );
    }

    renderChildren(renderable, routingProps) {
        if (renderable === null) {
            return React.createElement(this.props.loadingComponent, routingProps);
        } else if (!React.isValidElement(renderable)) {
            return React.createElement(renderable, routingProps);
        } else {
            return renderable;
        }
    }

    processChildren(unknown) {
        if (unknown instanceof Promise) {
            const instance = this;
            instance.promise = unknown;
            unknown.then(function handlePromiseResolution(result) {
                /**
                 * Is this promise still valid? If not, ignore the
                 * resolution.
                 */
                if (instance.promise === this) {
                    instance.setState({ children: result });
                    this.promise = null;
                }
            }.bind(unknown));

            return null;
        } else {
            return unknown;
        }
    }

    handleLocationChange = (/* event */) => {
        const currentValue = this.getContextValue(this.state.url, this.state.activeRoute);
        const { route: nextRoute, url: nextUrl } = findRoute(this.state.routes, Router.getURL());
        const nextValue = this.getContextValue(nextUrl, nextRoute);
        const result = this.props.routeWillChange(currentValue, nextValue);
        const cb = () => this.props.routeDidChange(nextValue, currentValue);
        const finish = () => this.recomputeRoutingState(nextUrl, this.state.routes, cb);

        // TODO: tests
        if (result === false) return;

        // TODO: tests
        else if (result instanceof Promise) result.then(() => finish());

        else finish();
    }

    recomputeRoutingState = (url, routes, cb) => {
        this.setState(
            this.getStateUpdateForUrl(
                url, routes === this.state.routes ? routes : this.processRoutes(routes)
            ), cb
        );
    }

    /**
     * Returns a react-router-esque object with parsed location and params.
     */
    getContextValue = (url, route) => {
        return {
            location: parseUrl(url),
            params: getRouteParamsForURL(route, url),
            route,
        };
    }

    processRoutes(routes) {
        if (process.env.NODE_ENV !== 'production') {
            if (!this.noFallbackWarningEmitted && routes.every(route => route.path !== '*')) {
                console.warn('no fallback route "*" was supplied. if a matching route is not found, the router will throw');
                this.noFallbackWarning = true;
            }
        }

        return routes.map(
            route =>
                Object.assign({}, route, {
                    params: extractParamsFromPath(route.path),
                    test: regexify(route.path),
                })
        );
    }

    getStateUpdateForUrl(url, routes) {
        const result = findRoute(routes, url);

        return {
            activeRoute: result.route,
            children: this.processChildren(
                result.route.render(
                    this.getContextValue(
                        result.url, result.route
                    )
                )
            ),
            routes,
            url: result.url,
        };
    }
}

/**
 * Compose it like this:
 *
 * <RoutingState>
 *   {({ location, params, route }) => {
 *      return <div>{location.pathname}</div>
 *   }}
 * </RoutingState>
 */
export const RoutingState = CONTEXT.Consumer;

/**
 * A polymorphic anchor link component. On click/tap/enter if the destination
 * matches a value route, the routing context will be modified without
 * reloading the page. Otherwise, it will act like a normal anchor link.
 *
 * If something other than an anchor tag is specified via props.as, a
 * [role="link"] attribute will be added for basic assistive technology support.
 *
 * Adds [data-active] if the given href matches the active route.
 */
export class Link extends React.PureComponent {
    static propTypes = {
        /**
         * An HTML tag name or valid ReactComponent class to be rendered. Must
         * be compatible with React.createElement.
         *
         * Defaults to an anchor "a" tag.
         */
        as: PropTypes.oneOfType([
            PropTypes.func,
            PropTypes.string,
        ]),

        /**
         * A valid relative or absolute URL string.
         */
        href: PropTypes.string.isRequired,

        /**
         * Any valid value of the anchor tag "target" attribute.
         *
         * See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-target
         */
        target: PropTypes.string,
    };

    static defaultProps = {
        as: 'a',
    };

    render() {
        return (
            <CONTEXT.Consumer>
                {({ route }) => {
                    return React.createElement(this.props.as, Object.assign({}, this.props, {
                        'data-active': valid(route.test, this.props.href) ? '' : undefined,
                        as: undefined,
                        href: this.shouldRenderAnchorProps() ? this.props.href : undefined,
                        role: this.shouldRenderRole() ? 'link' : undefined,
                        onClick: this.handleNavigationIntent,
                        onKeyDown: this.handleNavigationIntent,
                        onTouchEnd: this.handleNavigationIntent,
                        target: this.shouldRenderAnchorProps() ? this.props.target : undefined,
                    }));
                }}
            </CONTEXT.Consumer>
        );
    }

    shouldRenderAnchorProps() {
        return this.props.as === Link.defaultProps.as || typeof this.props.as !== 'string';
    }

    shouldRenderRole() {
        return this.props.as !== Link.defaultProps.as;
    }

    handleNavigationIntent = e => {
        if (e.type !== 'keydown' || (e.type === 'keydown' && (e.key === 'Enter' || e.key === 'Space'))) {
            e.preventDefault();
            e.stopPropagation();

            if (e.metaKey || e.target.getAttribute('target') === '_blank') {
                window.open(this.props.href);
            } else {
                route(this.props.href);
            }
        }
    };
}
