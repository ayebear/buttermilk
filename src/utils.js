import LiteURL from 'lite-url';

const PATH_EXTRACTION_R = /:[^/?#()]*/g;

// TODO: memoize?
export function extractParamsFromPath(path) {
    return (
        path.match(PATH_EXTRACTION_R) || []

    // remove the leading colon
    ).map(param => param.slice(1));
}

export function getRouteParamsForURL(route, url) {
    if (!route.params.length) return {};

    const result = url.match(route.test) || [];

    if (result.length) result.shift(); // discard the catchall result

    return result.reduce((params, result, index) => {
        return (params[route.params[index]] = result), params;
    }, {});
}

export function valid(validator, url) {
    if (validator instanceof RegExp) {
        return validator.test(url);
    } else if (validator instanceof Function) {
        return validator(url);
    }
}

export function findRoute(routes, url) {
    const route = routes.find(route => valid(route.test, url));

    if (route) {
        if (route.redirect) return findRoute(routes, route.redirect);
        else                return { route, url };
    }

    throw new Error(`could not find a route for url ${url}`);
}

export function getDisplayName(Component) {
    return Component.displayName || Component.name || 'Component';
}

export function parseUrl(url) {
    const parsed = new LiteURL(url);

    parsed.query = parsed.search.slice(1).split('&').reduce((params, pair) => {
        if (pair) {
            const idx = pair.indexOf('=');

            params[pair.slice(0, idx)] = pair.slice(idx + 1);
        }

        return params;
    }, {});

    return parsed;
}

/**
 * A client-side method for programmatically updating the routing state.
 *
 * Accepts a new url (absolute or relative) and an optional second boolean
 * parameter controlling if a new browser history entry should be created.
 *
 * If you want to change the routing state on server, just pass a new url to
 * the <Router> component.
 */
// TODO: tests
export function route(url, addNewHistoryEntry = true) {
    history[addNewHistoryEntry ? 'pushState' : 'replaceState']({}, '', url);

    // this is what triggers the routing to update
    window.dispatchEvent(new Event('popstate'));
}
