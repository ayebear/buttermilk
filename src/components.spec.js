import React from 'react';
import ReactDOM from 'react-dom';
import { Simulate } from 'react-dom/test-utils';

import { Link, Router, RoutingState } from './components';
import { route } from './utils';

let root;
const render = ({ url, ...props }) => {
    jest.spyOn(Router, 'getURL').mockImplementation(() => url);
    ReactDOM.render(<Router {...props} />, root)
};

beforeAll(() => document.body.appendChild((root = document.createElement('main'))));
beforeEach(() => jest.spyOn(console, 'warn').mockImplementation(() => {}));
afterEach(() => ReactDOM.unmountComponentAtNode(root));

describe('Router', () => {
    it('renders a simple element child', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <div>bar</div>,
            }],
            url: 'http://foo.com/foo',
        });

        expect(root.innerHTML).toContain('bar');
    });

    it('renders a child component class with the current routing state', () => {
        class Foo extends React.Component {
            render() {
                return JSON.stringify(this.props);
            }
        }

        render({
            routes: [{
                path: '/foo',
                render: () => Foo,
            }],
            url: 'http://foo.com/foo',
        });

        expect(JSON.parse(root.children[0].innerHTML)).toMatchObject({
            location: {
                href: 'http://foo.com/foo',
                params: {},
                pathname: '/foo',
                query: {},
            },
        });
    });

    it('warns if no fallback route was provided', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <div>bar</div>,
            }],
            url: 'http://foo.com/foo',
        });

        expect(console.warn).toHaveBeenCalled();
    });

    it('does not warn if a fallback route was provided', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <div>bar</div>,
            }, {
                path: '*',
                render: () => <div>oh well</div>,
            }],
            url: 'http://foo.com/foo',
        });

        expect(console.warn).not.toHaveBeenCalled();
    });

    it('handles popstate events', () => {
        const instance = render({
            routes: [{
                path: '/foo',
                render: () => <div>bar</div>,
            }, {
                path: '*',
                render: () => <div>oh well</div>,
            }],
            url: 'http://foo.com/foo',
        });

        expect(root.innerHTML).toContain('bar');

        Router.getURL.mockImplementation(() => 'http://foo.com/bar');
        window.dispatchEvent(new Event('popstate'));

        expect(root.innerHTML).toContain('oh well');
    });

    it('handles hashchange events', () => {
        render({
            routes: [{
                path: '#foo',
                render: () => <div>bar</div>,
            }, {
                path: '*',
                render: () => <div>oh well</div>,
            }],
            url: 'http://foo.com/#foo',
        });

        expect(root.innerHTML).toContain('bar');

        Router.getURL.mockImplementation(() => 'http://foo.com/');
        window.dispatchEvent(new Event('hashchange'));

        expect(root.innerHTML).toContain('oh well');
    });

    describe('routerDidInitialize', () => {
        it('is called upon construction with the initially-active routing', () => {
            const stub = jest.fn();

            render({
                routerDidInitialize: stub,
                routes: [{
                    path: '#foo',
                    render: () => <div>bar</div>,
                }, {
                    path: '*',
                    render: () => <div>oh well</div>,
                }],
                url: 'http://foo.com/#foo',
            });

            expect(stub).toHaveBeenCalledWith({
                location: expect.objectContaining({
                    hash: '#foo',
                }),
                params: {},
                route: expect.objectContaining({
                    path: '#foo'
                }),
            })
        });
    });

    describe('routeWillChange', () => {
        it('is called before applying a route update', () => {
            const stub = jest.fn();

            render({
                routes: [{
                    path: '/bar',
                    render: () => <div>bar</div>,
                }, {
                    path: '*',
                    render: () => <div>oh well</div>,
                }],
                routeWillChange: stub,
                url: 'http://foo.com/bar',
            });

            Router.getURL.mockImplementation(() => 'http://foo.com/');
            window.dispatchEvent(new Event('popstate'));

            expect(stub).toHaveBeenCalledWith(
                expect.objectContaining({
                    route: expect.objectContaining({
                        path: '/bar'
                    }),
                }),

                expect.objectContaining({
                    route: expect.objectContaining({
                        path: '*'
                    }),
                }),
            );
        });

        it('accepts a promise', async () => {
            const stub = jest.fn();

            render({
                routes: [{
                    path: '/bar',
                    render: () => <div>bar</div>,
                }, {
                    path: '*',
                    render: () => <div>oh well</div>,
                }],
                routeWillChange: () => Promise.resolve(),
                routeDidChange: stub,
                url: 'http://foo.com/bar',
            });

            Router.getURL.mockImplementation(() => 'http://foo.com/');
            await window.dispatchEvent(new Event('popstate'));

            expect(stub).toHaveBeenCalled();
        });

        it('is abortable by returning false', () => {
            const stub = jest.fn();

            render({
                routes: [{
                    path: '/bar',
                    render: () => <div>bar</div>,
                }, {
                    path: '*',
                    render: () => <div>oh well</div>,
                }],
                routeWillChange: () => false,
                routeDidChange: stub,
                url: 'http://foo.com/bar',
            });

            Router.getURL.mockImplementation(() => 'http://foo.com/');
            window.dispatchEvent(new Event('popstate'));

            expect(stub).not.toHaveBeenCalled();
        });

        it('is abortable by rejecting a promise', async () => {
            const stub = jest.fn();

            render({
                routes: [{
                    path: '/bar',
                    render: () => <div>bar</div>,
                }, {
                    path: '*',
                    render: () => <div>oh well</div>,
                }],
                routeWillChange: () => Promise.reject(),
                routeDidChange: stub,
                url: 'http://foo.com/bar',
            });

            Router.getURL.mockImplementation(() => 'http://foo.com/');
            await window.dispatchEvent(new Event('popstate'));

            expect(stub).not.toHaveBeenCalled();
        });
    });

    describe('routeDidChange', () => {
        it('is called after applying a route update', () => {
            const stub = jest.fn();

            render({
                routes: [{
                    path: '/bar',
                    render: () => <div>bar</div>,
                }, {
                    path: '*',
                    render: () => <div>oh well</div>,
                }],
                routeDidChange: stub,
                url: 'http://foo.com/bar',
            });

            Router.getURL.mockImplementation(() => 'http://foo.com/');
            window.dispatchEvent(new Event('popstate'));

            expect(stub).toHaveBeenCalledWith(
                expect.objectContaining({
                    route: expect.objectContaining({
                        path: '*'
                    }),
                }),

                expect.objectContaining({
                    route: expect.objectContaining({
                        path: '/bar'
                    }),
                }),
            );
        });
    });
});

describe('RoutingState', () => {
    it('inherits the routing state from the parent Router', () => {
        class Foo extends React.Component {
            render() {
                return (
                    <div id="inner">
                        <RoutingState>
                            {state => {
                                return JSON.stringify(state);
                            }}
                        </RoutingState>
                    </div>
                );
            }
        }

        render({
            routes: [{
                path: '/foo',
                render: () => Foo,
            }],
            url: 'http://foo.com/foo',
        });

        expect(JSON.parse(root.querySelector('#inner').innerHTML)).toMatchObject({
            location: {
                href: 'http://foo.com/foo',
                params: {},
                pathname: '/foo',
                query: {},
            },
        });
    });

    it('reflects updates in routing state', () => {
        class Foo extends React.Component {
            render() {
                return (
                    <div id="inner">
                        <RoutingState>
                            {state => {
                                return JSON.stringify(state);
                            }}
                        </RoutingState>
                    </div>
                );
            }
        }

        const instance = render({
            routes: [{
                path: '/foo(/:id)',
                render: () => Foo,
            }],
            url: 'http://foo.com/foo',
        });

        expect(JSON.parse(root.querySelector('#inner').innerHTML)).toMatchObject({
            location: {
                href: 'http://foo.com/foo',
                params: {},
                pathname: '/foo',
                query: {},
            },
        });

        Router.getURL.mockImplementation(() => 'http://foo.com/foo/bar');
        window.dispatchEvent(new Event('popstate'));

        expect(JSON.parse(root.querySelector('#inner').innerHTML)).toMatchObject({
            location: {
                href: 'http://foo.com/foo/bar',
                pathname: '/foo/bar',
                query: {},
            },
            params: {
                id: 'bar',
            },
        });
    });
});

describe('Link', () => {
    beforeEach(() => {
        jest.spyOn(history, 'pushState').mockImplementation(() => {});
        jest.spyOn(window, 'open').mockImplementation(() => {});
    });

    it('renders as an anchor tag by default', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <Link href="/bar">Baz</Link>,
            }],
            url: 'http://foo.com/foo',
        });

        expect(root.innerHTML).toMatchSnapshot();
    });

    it('renders as a different component if provided', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <Link as="div" href="/bar">Baz</Link>,
            }],
            url: 'http://foo.com/foo',
        });

        expect(root.innerHTML).toMatchSnapshot();
    });

    it('sets [data-active] if the link href matches the active route', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <Link href="/foo">Baz</Link>,
            }],
            url: 'http://foo.com/foo',
        });

        expect(root.innerHTML).toMatchSnapshot();
    });

    it('changes the routing state on click', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <Link href="/bar">Baz</Link>,
            }, {
                path: '/bar',
                render: () => 'It worked.',
            }],
            url: 'http://foo.com/foo',
        });

        Simulate.click(root.querySelector('a'));

        expect(history.pushState).toHaveBeenCalledWith({}, '', '/bar');
    });

    it('changes the routing state on touch end', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <Link href="/bar">Baz</Link>,
            }, {
                path: '/bar',
                render: () => 'It worked.',
            }],
            url: 'http://foo.com/foo',
        });

        Simulate.touchEnd(root.querySelector('a'));

        expect(history.pushState).toHaveBeenCalledWith({}, '', '/bar');
    });

    it('changes the routing state on enter press', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <Link href="/bar">Baz</Link>,
            }, {
                path: '/bar',
                render: () => 'It worked.',
            }],
            url: 'http://foo.com/foo',
        });

        Simulate.keyDown(root.querySelector('a'), {
            key: 'Enter',
        });

        expect(history.pushState).toHaveBeenCalledWith({}, '', '/bar');
    });

    it('changes the routing state on space press', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <Link href="/bar">Baz</Link>,
            }, {
                path: '/bar',
                render: () => 'It worked.',
            }],
            url: 'http://foo.com/foo',
        });

        Simulate.keyDown(root.querySelector('a'), {
            key: 'Space',
        });

        expect(history.pushState).toHaveBeenCalledWith({}, '', '/bar');
    });

    it('opens a new tab if metaKey is pressed while clicking the link', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <Link href="/bar">Baz</Link>,
            }, {
                path: '/bar',
                render: () => 'It worked.',
            }],
            url: 'http://foo.com/foo',
        });

        Simulate.click(root.querySelector('a'), {
            metaKey: true,
        });

        expect(window.open).toHaveBeenCalledWith('/bar');
        expect(history.pushState).not.toHaveBeenCalled();
    });

    it('opens a new tab if target is set to _blank', () => {
        render({
            routes: [{
                path: '/foo',
                render: () => <Link href="/bar" target="_blank">Baz</Link>,
            }, {
                path: '/bar',
                render: () => 'It worked.',
            }],
            url: 'http://foo.com/foo',
        });

        Simulate.click(root.querySelector('a'));

        expect(window.open).toHaveBeenCalledWith('/bar');
        expect(history.pushState).not.toHaveBeenCalled();
    });
});
