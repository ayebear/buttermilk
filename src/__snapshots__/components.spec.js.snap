// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders as a different component if provided 1`] = `"<div location=\\"[object Object]\\" params=\\"[object Object]\\" route=\\"[object Object]\\"><div role=\\"link\\">Baz</div></div>"`;

exports[`Link renders as an anchor tag by default 1`] = `"<div location=\\"[object Object]\\" params=\\"[object Object]\\" route=\\"[object Object]\\"><a href=\\"/bar\\">Baz</a></div>"`;

exports[`Link sets [data-active] if the link href matches the active route 1`] = `"<div location=\\"[object Object]\\" params=\\"[object Object]\\" route=\\"[object Object]\\"><a href=\\"/foo\\" data-active=\\"\\">Baz</a></div>"`;
